lib/pf/scan.pm

```diff
@@ -342,6 +343,59 @@ sub toString {
     return $self->{'_id'};
 }
 
+=head2 matchCategory
+
+Check if the category matches the configuration of the scanner
+
+=cut
+
+sub matchCategory {
+    my ($self, $node_attributes) = @_;
+    my $category = [split(/\s*,\s*/, $self->{_categories})];
+    my $node_cat = $node_attributes->{'category'};
+
+    get_logger->debug( sub { "Tring to match the role '$node_cat' against " . join(",", @$category) });
+    # validating that the node is under the proper category for provisioner
+    return @$category == 0 || any { $_ eq $node_cat } @$category;
+}
+
+=head2 matchOS
+
+Check if the OS matches the configuration of the scanner
+
+=cut
+
+sub matchOS {
+    my ($self, $node_attributes) = @_;
+    my @oses = @{$self->{_oses} || []};
+
+    my $fingerbank_info = pf::node::fingerbank_info($node_attributes->{mac}, $node_attributes);
+
+    get_logger->debug( sub { "Trying see if any of the devices : ".join(',', @{$fingerbank_info->{device_hierarchy_ids}})." are in : " . join(",", @oses) });
+    #if no oses are defined then it will match all the oses
+    return $TRUE if @oses == 0;
+    #if there is no device info then fail
+    return $FALSE if @{$fingerbank_info->{device_hierarchy_ids}} == 0;
+
+    foreach my $device_id (@{$fingerbank_info->{device_hierarchy_ids}}) {
+        if(any {$device_id eq $_} @oses){
+            return $TRUE;
+        }
+    }
+    return $FALSE;
+}
+
+=head2 match
+
+Check if the device matches the configuration of the scanner
+
+=cut
+
+sub match {
+    my ($self, $os, $node_attributes) = @_;
+    return $self->matchOS($node_attributes) && $self->matchCategory($node_attributes);
```

Comment:
change it to `$self->matchCategory($node_attributes) && $self->matchOS($node_attributes)` to avoid unnecessary look ups to fingerbank if matchCategory does not match.